kimchi/src/circuits/constraints.rs

```diff
@@ -373,25 +374,71 @@ impl<F: PrimeField, G: KimchiCurve<ScalarField = F>, OpeningProof: OpenProof<G>>
 impl<F: PrimeField> ConstraintSystem<F> {
     /// evaluate witness polynomials over domains
     pub fn evaluate(&self, w: &[DP<F>; COLUMNS], z: &DP<F>) -> WitnessOverDomains<F> {
-        // compute shifted witness polynomials
-        let w8: [E<F, D<F>>; COLUMNS] =
-            array::from_fn(|i| w[i].evaluate_over_domain_by_ref(self.domain.d8));
-        let z8 = z.evaluate_over_domain_by_ref(self.domain.d8);
-
-        let w4: [E<F, D<F>>; COLUMNS] = array::from_fn(|i| {
-            E::<F, D<F>>::from_vec_and_domain(
-                (0..self.domain.d4.size)
-                    .map(|j| w8[i].evals[2 * j as usize])
-                    .collect(),
-                self.domain.d4,
-            )
-        });
+        // the idea is to have threading minimised below a certain threshold.
+        let min_len = {
+            let threads_to_use = if w[0].len() <= 2048 {
+                1
+            } else {
+                rayon::max_num_threads()
+            };
+            // min batch size is COLUMNS, when threads_to_use == 1,
+            // in which case every iterator will use one thread
+            // min batch size is COLUMNS / N where N = max_num_threads otherwise
+            // an extra max is to ensure the result is not 0
+            std::cmp::max(1, COLUMNS / threads_to_use)
+        };
+
+        // compute shifted witness polynomials and z8, all in parallel
+        let (w8, z8): ([E<F, D<F>>; COLUMNS], _) = {
+            let mut res = w
+                .par_iter()
+                .chain(rayon::iter::once(z))
+                .with_min_len(min_len)
+                .map(|elem| elem.evaluate_over_domain_by_ref(self.domain.d8))
+                .collect::<Vec<_>>();
+            let z8 = res[COLUMNS].clone();
```

Comment:
Is it worth doing this? It implies an unnecessary cloning + truncate, for, I think, a negligible gain to evaluate z in parallel with the columns.
Cloning will allocate new memory and truncate will call an unnecessary drop_in_place. On the web-assembly side, it might not be negligible.
What about a `rayon.join`? Or simply a sequential call to `w.par_iter()` without `chain(rayon::iter::once(z))` followed by a simple call to `z.evaluate_over_domain_by_ref(self.domain.d8)`.

Reply:
It seems yes, as cloning + deallocating does not seem to cost much.
Makes sense!